### PR TITLE
Qt: Do not require mkDerivation

### PR DESCRIPTION
**Pull request by ttuegel**
###### Motivation for this change

This removes the need to use a custom `mkDerivation` with Qt applications. For compatibility, it remains, but it doesn't do much besides include `wrapQtAppsHook`.

See also: https://github.com/NixOS/nixpkgs/issues/65399#issuecomment-541076520

TODO:

- [ ] Port #70691.
- [ ] Update documentation to remove references to `mkDerivation`.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

**Comment by worldofpeace** [comment]
@ttuegel By port https://github.com/NixOS/nixpkgs/pull/70691 you mean qtbase setup hook checks for the presence of wrapQtAppsHook?

**Comment by ttuegel** [comment]
> By port #70691 you mean qtbase setup hook checks for the presence of wrapQtAppsHook?

@worldofpeace Something like that. I'm not sure what the best option is: check for presence, or include by default. I like the idea of doing the right thing by default (including wrapQtAppsHook), but I know that trying to be "magic" often makes things difficult to understand. On the other hand, it is always possible to set `dontWrapQtApps` to disable the magic.

**Comment by doronbehar** [comment]
How is the debug flag related to `mkDerivation`?

**Comment by ttuegel** [comment]
> How is the debug flag related to `mkDerivation`?

We have to set certain `CFLAGS` in all dependencies, based on the `debug` flag.

**Review by worldofpeace — APPROVED** [review]
Let's try it out in staging

**Review by SuperSandro2000 — APPROVED** [review]
Lets go to staging! 🎰

File: pkgs/development/libraries/qt-5/hooks/qtbase-setup-hook.sh [hunk 1]
```diff
@@ -1,10 +1,35 @@
...
+# Disable debug symbols if qtbase was built without debugging.
+# This stops -dev paths from leaking into other outputs.
+if [ -z "@debug@" ]; then
+    NIX_CFLAGS_COMPILE="${NIX_CFLAGS_COMPILE}${NIX_CFLAGS_COMPILE:+ }-DQT_NO_DEBUG"
+fi
+
+# Integration with CMake:
+# Set the CMake build type corresponding to how qtbase was built.
+if [ -n "@debug@" ]; then
+    cmakeFlags="${cmakeFlags-}${cmakeFlags:+ }-DCMAKE_BUILD_TYPE=Debug"
```
Comment (jtojnar) [inline pkgs/development/libraries/qt-5/hooks/qtbase-setup-hook.sh:28]: Maybe use `cmakeBuildType` instead?

File: pkgs/development/libraries/qt-5/mkDerivation.nix [hunk 2]
```diff
@@ -9,19 +9,6 @@ args:
...
-      optional (!debug) "-DQT_NO_DEBUG"
-      ++ lib.toList (args.NIX_CFLAGS_COMPILE or []));
-
-    cmakeFlags =
-      (args.cmakeFlags or [])
-      ++ [
-        ("-DCMAKE_BUILD_TYPE=" + (if debug then "Debug" else "Release"))
-      ];
-
     enableParallelBuilding = args.enableParallelBuilding or true;
```
Comment (jtojnar) [inline pkgs/development/libraries/qt-5/mkDerivation.nix:12]: This should be redundant as both cmake and qmake setup hooks already enable this by default.

File: pkgs/development/libraries/qt-5/hooks/qtbase-setup-hook.sh (hunk 1 above)
Comment (ttuegel) [reply pkgs/development/libraries/qt-5/hooks/qtbase-setup-hook.sh:28]: :+1: It is... possible that did not exist when this pull request was originally drafted. :laughing:

**Comment by ttuegel** [comment]
I am closing this pull request because my branch is completely broken and I don't know why. I will re-open later.

**Pull request closed without merging**